ui/src/shared/utils/useVisDomainSettings.ts

```diff
@@ -54,3 +54,42 @@ export const useVisDomainSettings = (
 
   return [domain, setDomain, resetDomain]
 }
+
+export const getRemainingRange = (
+  data: NumericColumnData = [],
+  timeRange: TimeRange | null,
+  storedDomain: number[]
+) => {
+  const range = extent((data as number[]) || [])
+  if (range && range.length >= 2) {
+    const startTime = getStartTime(timeRange)
+    const endTime = getEndTime(timeRange)
+    const start = storedDomain[0]
+      ? storedDomain[0]
+      : Math.min(startTime, range[0])
+    const end = storedDomain[1] ? storedDomain[1] : Math.max(endTime, range[1])
+    return [start, end]
+  }
+  return range
+}
+
+export const useVisYDomainSettings = (
+  storedDomain: number[],
+  data: NumericColumnData,
+  timeRange: TimeRange | null = null
+) => {
+  const initialDomain = useMemo(() => {
+    if (storedDomain === null || storedDomain.every(val => val === null)) {
+      return getValidRange(data, timeRange)
+    }
+    if (storedDomain.includes(null)) {
+      return getRemainingRange(data, timeRange, storedDomain)
+    }
```

Comment:
this code loops over the same structure twice to perform essentially the same check (is the current item null). this happens while rendering. it's good you're using memoization, but you could write this logic using a single loop. what kind of array sizes are we dealing with in these loops?

Reply:
I'm not sure I follow you, would you mind elaborating on this a bit?